### PR TITLE
add tokenreviews endpoint to implement webhook

### PR DESCRIPTION
**Pull request by deads2k**
Wires up an API resource under `apis/authentication.k8s.io/v1beta1` to expose the webhook token authentication API as an API resource.  This allows one API server to use another for authentication and uses existing policy engines for the "authoritative" API server to controller access to the endpoint.

@cjcullen you wrote the initial type

### CODE REVIEW

**Comment by deads2k** [comment]
The package had to move to make client generation work, which need to be generate to have kubectl available to call create.

File: pkg/registry/tokenreview/storage.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	"k8s.io/kubernetes/pkg/api"
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+	"k8s.io/kubernetes/pkg/auth/authenticator"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type REST struct {
+	tokenAuthenticator authenticator.Request
```
Comment (deads2k) [inline pkg/registry/tokenreview/storage.go]: @kubernetes/kube-iam I was torn about this.  We have an unknown number of token authenticators that are included in the Authentication chain.  I could plumb through a set of union token authentications down through the masterconfig that are required to operate the same way as the authentication chain or I could create a fake request that only has a bearer token and count on the existing chain to do the right thing.

I chose to make the fake request, figuring it was easier to get right.  My only concern is about whether some authenticator does weird things with requests it doesn't recognize.

**Comment by thockin** [comment]
Out of curiosity, why odes this *one* API have "k8s.io" in the name?

**Comment by deads2k** [comment]
> Out of curiosity, why odes this one API have "k8s.io" in the name?

API groups were supposed to dns names so that they'd namespace nicely (like a java package) with prioritized completion for the CLI tooling.  We have `authorization.k8s.io`, `rbac.authorization.k8s.io`, and `authentication.k8s.io`.  See comment here: https://github.com/kubernetes/kubernetes/pull/24902#discussion_r62391331

The package folders were going to follow that, but there were some generator problems that were easier to resolve by eliminating the dots.

**Comment by deads2k** [comment]
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/26760

**Comment by erictune** [comment]
@kubernetes/sig-cluster-federation 
This will be useful for delegating auth from the federation apiserver to the main apiserver of the cluster that hosts it.

**Comment by deads2k** [comment]
@smarterclayton I need help with protobuf, tests complained about not having it, but the generator seems decidedly unhappy about `map[string][]string`

**Comment by smarterclayton** [comment]
Yes there's an issue open.  The optional protobuf work supports it - we
need to convert []String to a type, then use the "optional" field to handle
whether it's set.   Cannot generate protobuf for map[string][]string,
blocks authorization from being generated
<https://github.com/kubernetes/kubernetes/issues/27259>

**Comment by deads2k** [comment]
@bgrant0607 Looking for API approval.

The type existed in our external contracts before and was exposed as our authentication webhook type.  This wires up our API server to fulfill the endpoint contract and enables authn delegation as @erictune noted.

**Comment by deads2k** [comment]
@kubernetes/api-review-team This exposes an already existing API type. ptal.

**Comment by deads2k** [comment]
split into two commits to keep boring changes from drowning out interesting ones.

File: pkg/master/master.go [hunk 2]
```diff
@@ -384,6 +386,26 @@ func (m *Master) InstallAPIs(c *Config) {
 		apiGroupsInfo = append(apiGroupsInfo, apiGroupInfo)
 	}
 
+	if c.APIResourceConfigSource.AnyResourcesForVersionEnabled(authenticationv1beta1.SchemeGroupVersion) {
+		resources := m.getAuthenticationResources(c)
+		groupMeta := registered.GroupOrDie(authenticationv1beta1.GroupName)
+
+		// Hard code preferred group version to apps/v1alpha1
```
Comment (erictune) [inline pkg/master/master.go]: s/apps/authentication/

**Comment by erictune** [comment]
This looks good to me.  I'd like @cjcullen  to sign off too.

**Comment by erictune** [comment]
@deads2k what RBAC roles would you think should have permissions to use this endpoint?

**Comment by bgrant0607** [comment]
@deads2k I delegate API approval to @erictune

**Comment by bgrant0607** [comment]
@deads2k Tests failed.

**Comment by cjcullen** [comment]
Looks fine to me. We do need to think about who gets to access this API (as well as the SubjectAccessReviews API).

**Comment by deads2k** [comment]
> @deads2k what RBAC roles would you think should have permissions to use this endpoint?

In openshift, we're going to open tokenreviews up to nodes and clusteradmins.  In kube, I suspect you'll create a "federated-server" role which will encompass this and SAR, to handle delegated auth cases.

I think in the end, we'll have a separate node authorizer based on ref counting summaries of the logical resource graph.

**Comment by deads2k** [comment]
Looks like the generator changed on me again, I'll update tomorrow morning.

File: hack/update-swagger-spec.sh [hunk 3]
```diff
@@ -74,7 +74,7 @@ APISERVER_PID=$!
 kube::util::wait_for_url "http://127.0.0.1:${API_PORT}/healthz" "apiserver: "
 
 SWAGGER_API_PATH="http://127.0.0.1:${API_PORT}/swaggerapi/"
-DEFAULT_GROUP_VERSIONS="v1 autoscaling/v1 batch/v1 batch/v2alpha1 extensions/v1beta1 apps/v1alpha1 policy/v1alpha1 rbac.authorization.k8s.io/v1alpha1 certificates/v1alpha1"
+DEFAULT_GROUP_VERSIONS="v1 autoscaling/v1 authentication.k8s.io/v1beta1 batch/v1 batch/v2alpha1 extensions/v1beta1 apps/v1alpha1 policy/v1alpha1 rbac.authorization.k8s.io/v1alpha1 certificates/v1alpha1"
```
Comment (sttts) [inline hack/update-swagger-spec.sh]: nit: not alphabetically. auto > auth

File: pkg/apis/authentication/types.go [hunk 4]
```diff
@@ -38,11 +46,14 @@ type TokenReviewSpec struct {
...
 
 // TokenReviewStatus is the result of the token authentication request.
+// This type mirrors the authentication.Token interface
 type TokenReviewStatus struct {
 	// Authenticated indicates that the token was associated with a known user.
 	Authenticated bool
 	// User is the UserInfo associated with the provided token.
 	User UserInfo
+	// Error indicates that the token couldn't be checked
+	Error string
```
Comment (sttts) [inline pkg/apis/authentication/types.go]: no omitempty?

File: pkg/apis/authentication/types.go [hunk 5]
```diff
@@ -57,5 +68,8 @@ type UserInfo struct {
 	// The names of groups this user is a part of.
 	Groups []string
 	// Any additional information provided by the authenticator.
-	Extra map[string][]string
+	Extra map[string]ExtraValue
```
Comment (sttts) [inline pkg/apis/authentication/types.go]: omitempty?

File: pkg/apis/authentication/types.go [hunk 6]
```diff
@@ -17,12 +17,20 @@ limitations under the License.
...
+// +genclient=true
+// +nonNamespaced=true
+// +noMethods=true
+
 // TokenReview attempts to authenticate a token to a known user.
 type TokenReview struct {
 	unversioned.TypeMeta
+	// ObjectMeta fulfills the meta.ObjectMetaAccessor interface so that the stock
+	// REST handler paths work
+	api.ObjectMeta `json:"metadata,omitempty"`
```
Comment (sttts) [inline pkg/apis/authentication/types.go]: looks like the internal representation has not json annotation at all. This is the only one.

File: pkg/client/unversioned/authentication.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+func NewAuthenticationOrDie(c *restclient.Config) *AuthenticationClient {
+	client, err := NewAuthentication(c)
+	if err != nil {
+		panic(err)
+	}
+	return client
+}
+
+func setAuthenticationDefaults(config *restclient.Config) error {
+	// if authentication group is not registered, return an error
```
Comment (sttts) [inline pkg/client/unversioned/authentication.go]: comment doesn't add anything that the code below tells us.

File: pkg/client/unversioned/helper.go [hunk 8]
```diff
@@ -123,7 +133,18 @@ func New(c *restclient.Config) (*Client, error) {
 		}
 	}
 
-	return &Client{RESTClient: client, AutoscalingClient: autoscalingClient, BatchClient: batchClient, CertificatesClient: certsClient, ExtensionsClient: extensionsClient, DiscoveryClient: discoveryClient, AppsClient: appsClient, PolicyClient: policyClient, RbacClient: rbacClient}, nil
+	return &Client{
+		RESTClient:           client,
+		AppsClient:           appsClient,
+		AutoscalingClient:    autoscalingClient,
```
Comment (sttts) [inline pkg/client/unversioned/helper.go]: order

File: pkg/master/master.go (hunk 2 above)
Comment (sttts) [inline pkg/master/master.go]: s/alpha/beta/

File: pkg/master/master.go [hunk 9]
```diff
@@ -384,6 +386,26 @@ func (m *Master) InstallAPIs(c *Config) {
 		apiGroupsInfo = append(apiGroupsInfo, apiGroupInfo)
 	}
 
+	if c.APIResourceConfigSource.AnyResourcesForVersionEnabled(authenticationv1beta1.SchemeGroupVersion) {
+		resources := m.getAuthenticationResources(c)
```
Comment (sttts) [inline pkg/master/master.go]: nit: following the convention of the the other scheme group above and below: `authenticationResources`

File: pkg/apis/authentication/types.go (hunk 4 above)
Comment (deads2k) [inline pkg/apis/authentication/types.go]: > no omitempty?

no internal json tags.

File: pkg/client/unversioned/authentication.go (hunk 7 above)
Comment (deads2k) [inline pkg/client/unversioned/authentication.go]: > comment doesn't add anything that the code below tells us.

same template as the rest.

File: pkg/master/master.go (hunk 9 above)
Comment (deads2k) [inline pkg/master/master.go]: > nit: following the convention of the the other scheme groups above and below: authenticationResources

Bad convention.  Also, it all goes away in https://github.com/kubernetes/kubernetes/pull/27554

File: pkg/registry/tokenreview/storage.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+func NewREST(tokenAuthenticator authenticator.Request) *REST {
+	return &REST{tokenAuthenticator: tokenAuthenticator}
+}
+
+func (r *REST) New() runtime.Object {
+	return &authentication.TokenReview{}
+}
+
+// Create registers a given new ResourceAccessReview instance to r.registry.
```
Comment (sttts) [inline pkg/registry/tokenreview/storage.go]: s/ResourceAccessReview/TokenReview/

File: pkg/registry/tokenreview/storage.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+		tokenReview.Status.Error = err.Error()
+	}
+	if tokenUser != nil {
+		tokenReview.Status.User = authentication.UserInfo{
+			Username: tokenUser.GetName(),
+			UID:      tokenUser.GetUID(),
+			Groups:   tokenUser.GetGroups(),
+		}
+		for k, v := range tokenUser.GetExtra() {
+			tokenReview.Status.User.Extra[k] = authentication.ExtraValue(v)
```
Comment (sttts) [inline pkg/registry/tokenreview/storage.go]: `tokenReview.Status.User.Extra` is nil

**Comment by deads2k** [comment]
These generators are going to be the death of me.

@smarterclayton Any idea about this one?

```
# k8s.io/kubernetes/pkg/apis/authentication/v1beta1
pkg/apis/authentication/v1beta1/generated.pb.go:384: this.Items undefined (type *ExtraValue has no field or method Items)
!!! Error in /go/src/k8s.io/kubernetes/hack/lib/golang.sh:501
  'go install "${goflags[@]:+${goflags[@]}}" -ldflags "${goldflags}" "${nonstatics[@]:+${nonstatics[@]}}"' exited with status 2
Call stack:
  1: /go/src/k8s.io/kubernetes/hack/lib/golang.sh:501 kube::golang::build_binaries_for_platform(...)
  2: /go/src/k8s.io/kubernetes/hack/lib/golang.sh:670 kube::golang::build_binaries(...)
  3: hack/make-rules/build.sh:26 main(...)
```

**Comment by deads2k** [comment]
comments addressed, now with happy generated code.

**Comment by erictune** [comment]
CJ okayed via side channel.

**Comment by erictune** [comment]
API looks good.

LGTM

**Comment by erictune** [comment]
@stts thank you for the careful code review.

**Comment by wojtek-t** [comment]
It seems that this PR broke GKE. After this PR was merged, I'm seeing a bunch of logs like these:
```
23:23:00 Created [https://test-container.sandbox.googleapis.com/v1/projects/k8s-jkns-e2e-gke-ci/zones/us-central1-f/clusters/jenkins-e2e].
23:23:00 kubeconfig entry generated for jenkins-e2e.
23:23:00 NAME         ZONE           MASTER_VERSION                    MASTER_IP       MACHINE_TYPE   NODE_VERSION                      NUM_NODES  STATUS
23:23:00 jenkins-e2e  us-central1-f  1.4.0-alpha.1.564+8ead63f1270455  146.148.90.159  n1-standard-2  1.4.0-alpha.1.564+8ead63f1270455  3          RUNNING
23:23:00 ... calling validate-cluster
23:23:01 failed to find client for version v1: error: You must be logged in to the server (the server has asked for the client to provide credentials)
23:23:01 (kubectl failed, will retry 2 times)
23:23:02 failed to find client for version v1: error: You must be logged in to the server (the server has asked for the client to provide credentials)
23:23:02 (kubectl failed, will retry 1 times)
```

I'm going to revert this one to fix the merge-queue.

**Comment by piosz** [comment]
The interesting thing here is that GKE smoke tests passed. There might be some issue with testing infrastructure.

@kubernetes/test-infra-maintainers

**Comment by wojtek-t** [comment]
FYI - reverting the PR helped.

It seems to me that GKE smoke tests are not catching what they are supposed to catch. What are they actually doing?

**Comment by wojtek-t** [comment]
@kubernetes/test-infra-maintainers @kubernetes/goog-gke - since it was failing in verify-cluster, it clearly means that GKE smoke tests are configured differently than all other suites. This should be fixed.

File: pkg/apis/authentication/types.go [hunk 12]
```diff
@@ -17,12 +17,20 @@ limitations under the License.
...
+// +genclient=true
+// +nonNamespaced=true
+// +noMethods=true
+
 // TokenReview attempts to authenticate a token to a known user.
 type TokenReview struct {
 	unversioned.TypeMeta
+	// ObjectMeta fulfills the meta.ObjectMetaAccessor interface so that the stock
+	// REST handler paths work
+	api.ObjectMeta
```
Comment (cjcullen) [inline pkg/apis/authentication/types.go]: I don't have the logs to prove it, but I think that these additional fields are choking up our deserialization on the GKE server side. I'll see what we can do about that.

Comment (deads2k) [inline pkg/apis/authentication/types.go]: > I don't have the logs to prove it, but I think that these additional fields are choking up our deserialization on the GKE server side. I'll see what we can do about that.

Thanks, let me know?

Comment (cjcullen) [inline pkg/apis/authentication/types.go]: Confirmed that we do strict field checking on our side. I either need to figure out if we can turn that off, or add all of the fields from ObjectMeta to our server-side TokenReview representation.

Why does this PR add api.ObjectMeta to the TokenReview, but #20573 doesn't add it to SubjectAccessReview?

Comment (deads2k) [inline pkg/apis/authentication/types.go]: > Why does this PR add api.ObjectMeta to the TokenReview, but #20573 doesn't add it to SubjectAccessReview?

When I did the SAR long ago it wasn't used.  I'll probably have to add it there too.

Comment (liggitt) [inline pkg/apis/authentication/types.go]: @cjcullen any word on making the token review more tolerant?

**Comment by ingvagabund** [comment]
@deads2k not every file under ``pkg/client/clientset_generated`` was generated automatically, right? There are at least two that I believe were hand-crafted:
* pkg/client/clientset_generated/internalclientset/typed/authentication/unversioned/fake/fake_generated_expansion.go 
* pkg/client/clientset_generated/internalclientset/typed/authentication/unversioned/generated_expansion.go 

Or is there another command besides ``./hack/update-codegen.sh`` that needs to be run?

**Comment by deads2k** [comment]
> I can open an issue on that if it makes sense or there is not already one for "not every file under clientset_generated is generated".

The expansions are there for the specific purpose of being hand crafted.

**Comment by caesarxuchao** [comment]
Right. Will comments in expansion files and doc.go be enough to make clear these are not generated? I can send a PR to do that.

**Comment by ingvagabund** [comment]
@deads2k thanks. I can confirm that by observation. Found 21 files under ``clientset_generated`` ending with ``_expansion.go`` that I needed to modify. 21 is still ok, though it can get annoying to do regularly if the number start to grow over time.

@caesarxuchao that would be great. Could the generator (``./hack/update-codegen.sh``) also report that? E.g. find all files under affected ``clientset_generated`` directories and list them with warning all of them are possible candidates for manual update.

**Pull request merged**